[DS-3940] SHERPA API v2 refactor

kshepherd opened the pull request:
Developed by The Library Code with the support of Technische Universität Hamburg

## References
* Fixes #7287
* [JIRA issue link](https://jira.lyrasis.org/browse/DS-3940) 
* [Register for SHERPA v2 API key here](https://v2.sherpa.ac.uk/romeo/api.html)

## Description
The legacy SHERPA ROMeO API that DSpace 6.x (and earlier) is using in both XMLUI and JSPUI for journal deposit policy lookup will be retired in April. 
This refactor uses the new v2 API.
Some changes to display are necessary because of the way policy, journal and publisher data is structured in the new API.

## Instructions for Reviewers

To test this PR:
* Register for a SHERPA API v2 key and configure it in local.cfg
* Begin a submission with a valid ISSN entered into metadata
* At Upload Step, inspect the SHERPA ROMeO policy information and compare it to the source policy page, existing DSpace legacy API response, common sense, etc.
* Try with variant options -- no ISSN, multiple ISSNs, invalid ISSN, mix of valid and invalid ISSNs

List of changes in this PR:
* Deprecated old SHERPA* beans and moved them to org.dspace.app.sherpa.v1
* Wrote new SHERPAResponse class and the beans to represent new data classes in org.dspace.app.sherpa.v2
* Made changes to the servlet and JSP display in JSPUI where appropriate - notably, no more colours, and conditions are grouped by version and option rather than lumped together in "general conditions"
* Made small changes to XMLUI - no visible difference because it just provides an indicator and a link to the ROMeO policy page
* As the v2 API filters don't allow searching for multiple ISSNs, made a change so that just the first valid API response for a set of ISSNs would be used

kshepherd committed:
[DS-3940] SHERPA API v2 refactor

kshepherd committed:
[DS-3940] Apply missing license header for one bean class

nwoodward commented:
This PR looks really good. I have a few observations and a question. On both the XMLUI and JSPUI submission forms I can enter in the same ISSN multiple times in the Describe Step. If I'm in XMLUI then on the Upload Step I see multiple instances of the same journal in Publisher Information. If I'm in JSPUI, in the Upload Step I only see the journal information for the first ISSN in the list, even if there are multiple valid ISSNs in the Describe Step. 

Is that how things are supposed to work? I checked on a DSpace with the old SHERPA API enabled, and I was able to add the same ISSN multiple times in the Describe Step, so I'm thinking this is probably intentional.

kshepherd commented:
Thanks @nwoodward 
I haven't changed the way XMLUI works for multiple ISSNs - though I did for JSPUI to just take the first, as you note (the UI approach isn't really set up to display multiple results).

The code comments for XMLUI indicates that displaying indicators/links for *each* journal/ISSN was intentional, though probably not for duplicate ISSNs! That should be a small easy fix.

kshepherd committed:
[DS-3940] Revert "getISSNs" to return LinkedHashSet to avoid duplicates

kshepherd commented:
> This PR looks really good. I have a few observations and a question. On both the XMLUI and JSPUI submission forms I can enter in the same ISSN multiple times in the Describe Step. If I'm in XMLUI then on the Upload Step I see multiple instances of the same journal in Publisher Information. If I'm in JSPUI, in the Upload Step I only see the journal information for the first ISSN in the list, even if there are multiple valid ISSNs in the Describe Step.
> 
> Is that how things are supposed to work? I checked on a DSpace with the old SHERPA API enabled, and I was able to add the same ISSN multiple times in the Describe Step, so I'm thinking this is probably intentional.

This is now fixed - in my changes, I thought I was simplifying some code by returning a List instead of a Set for the getISSNs method, but it turns out LinkedHashSet was being used specifically to avoid duplicates.

nwoodward reviewed (approved):
:+1: Everything looks good!

J4bbi reviewed (commented):
I think it would be good if a summary could be provided on the upload step page.
Something like "SHERPA gave 4 results out of 5 ISSNs entered".

Especially on the JSPUI there should be an explanation that the result is only for the first ISSN.

It is not intuitive that if two ISSNs are provided, the first one bogus and the second okay, that a "No results" message is given without any caveat.

kshepherd committed:
[DS-3940] Changes to handle multiple ISSN queries / errors in JSPUI

kshepherd commented on dspace-jspui/src/main/webapp/sherpa/css/sherpa.css:
```diff
@@ -25,16 +25,35 @@ div.policy div p img {margin: 10px;display: inline-block;}
-.sherpaBox img {
+#sherpaBox {
+	max-height: 50em;
+	overflow: scroll;
```
Thanks, I've changed the overflow to auto

kshepherd commented on dspace-api/src/main/resources/Messages.properties:
```diff
@@ -901,17 +901,17 @@ jsp.search.filter.original_bundle_filenames                      = Filename
+jsp.sherpa.error = <p class="sherpaError">Sorry, we have had trouble querying the SHERPA/RoMEO Database. No data is available, try later or check directly at the <a href="https://v2.sherpa.ac.uk/romeo/" target="_blank">SHERPA/RoMEO WebSite</a>.</p>
+jsp.sherpa.noresult = <p class="sherpaNoResult">Sorry, there is no data in the <a href="http://www.sherpa.ac.uk/romeo.php" target="_blank">SHERPA/RoMEO Database</a> for the ISSNs that you have entered.</p>
+jsp.sherpa.oneresult = <p>The <a href="https://v2.sherpa.ac.uk/romeo/" target="_blank">SHERPA/RoMEO</a> database provides the following data for the journal that you have entered.</p>
 jsp.sherpa.moreresults = <p>The ISSNs that you have entered match with multiple journals, please review them. You can find the publisher policy for these journals below.</p>
```
I have changed the JSPUI handling so that each ISSN is now search separately, as with XMLUI. Additional results will be displayed below (since we are scrolling overflow, this is generally OK), as well as adding messages to indicate when there are multiple results, and/or when at least one ISSN returned an error or invalid response

kshepherd committed:
[DS-3940] Remove unused imports

nwoodward commented:
The additional results display in JSPUI looks good, and I saw the some-errors message as well.

J4bbi commented:
The multiple issn handling in JSPUI looks really good. Nice work.

KevinVdV reviewed (changes requested):
@kshepherd I made a review of the code, there are a few things below that I would change.

I also noticed that not everything is parsed yet (which is fine), but it would be best to store the publisher url as it is used in the xmlui, currently the publisher url is linking to the DSpace url on that page.

kshepherd committed:
[DS-3940] Remove formatted text from response object as per review

kshepherd committed:
[DS-3940] Close input stream reader, throw IOException from parse methods

kshepherd committed:
[DS-3940] Remove ROMeO colour from data models, as per review

kshepherd committed:
[DS-3940] Split large parseJSON method into functional parts

including slightly restructuring Journal and Permitted Version
to work better with the way things are scoped in the new structure

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java:
```diff
@@ -0,0 +1,471 @@
+     * @param jsonData
+     */
+    private void parseJSON(InputStream jsonData) {
+        JSONTokener jsonTokener = new JSONTokener(new InputStreamReader(jsonData));
```
Thanks, I have done this in a finally block (and the main methods now throw IOException)

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:
```diff
@@ -89,7 +104,7 @@ public SHERPAResponse searchByJournalISSN(String query)
 
                 if (null != responseBody)
-                    sherpaResponse = new SHERPAResponse(responseBody.getContent());
+                    sherpaResponse = new SHERPAResponse(responseBody.getContent(), SHERPAResponse.SHERPAFormat.JSON);
```
Thanks, I have done this in a finally block (and the main methods now throw IOException)

KevinVdV replied:
@kshepherd Did you forget to commit the fix here ? Because this class wasn't altered recently.

kshepherd replied:
Thanks, I've pushed a change now

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java:
```diff
@@ -0,0 +1,471 @@
+                            }
+                        }
+                        if (item.has("url")) {
+                            this.journalUrl = item.getString("url");
```
Removed

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAPublisher.java:
```diff
@@ -0,0 +1,92 @@
+
+    // This is deprecated - we should not rely on 'SHERPA ROMeO colour' anymore and it is a bit dangerous
+    // to try and calculated it ourselves, now that it is no longer supplied
+    private String romeoColour;
```
Removed

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java:
```diff
@@ -0,0 +1,471 @@
+     * Parse the SHERPA v2 API JSON and construct Romeo policy data for display
+     * @param jsonData
+     */
+    private void parseJSON(InputStream jsonData) {
```
I've split this method into parts, so there are now separate methods for parsing PublisherPolicy, PermittedVersion, Publisher, Journal, and SystemMetadata. I have had to slightly rework some other logic and models to handle the change in scope, but it is working the same and hopefully is easier to follow. I also added javadoc annotation for each method.

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java:
```diff
@@ -0,0 +1,471 @@
+                    String licenceUrl = I18nUtil.getMessage("jsp.sherpa.license-url");
+                    this.license = licenceText;
+                    this.licenseURL = licenceUrl;
+                    this.disclaimer = disclaimerText;
```
Removed

kshepherd committed:
[DS-3940] Ensure publisher URL is set correctly, refactor parse pub name

kshepherd commented:
> @kshepherd I made a review of the code, there are a few things below that I would change.
> 
> I also noticed that not everything is parsed yet (which is fine), but it would be best to store the publisher url as it is used in the xmlui, currently the publisher url is linking to the DSpace url on that page.

Hi @KevinVdV , thanks for the review and change requests - I have made all the changes requested

- Some unused and obsolete attributes were removed
- The formatted text (not part of actual response object) was removed
- the parseJSON method was split into several methods to make everything a bit easier to follow and maintain
- javadocs were added
- publisher URL is fixed
- inputstreamreader is always closed

Could you please take another look and update your review? Thanks!

kshepherd commented:
(note, the Travis CI build appears to be failing because of a Forbidden response from a mvn repo - I'll restart it periodically)

kshepherd commented:
closing and reopening to force rebuild

KevinVdV reviewed (changes requested):
@kshepherd Just one more thing that needs to be fixed, you mentioned that you did fix it, so perhaps you forgot to commit it ?

kshepherd committed:
[DS-3940] Explicitly close SHERPAService response content InputStream

KevinVdV commented on dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:
```diff
@@ -103,10 +106,14 @@ public SHERPAResponse searchByJournalISSN(String query)
+                if (null != responseBody) {
+                    InputStream content = responseBody.getContent();
+                    sherpaResponse = new SHERPAResponse(content, SHERPAResponse.SHERPAFormat.JSON);
+                    content.close();
```
Can we place this content.close() in a finally ? If for whatever reason the parsing fails the inputStream won't be closed. After this change I approve of this PR !

Ps: We should also remember to port this PR to the latest DSpace 7 master.

kshepherd replied:
Good point, I did that in my other implementation, I'm not sure why this one was different.
I have rewritten this small section for a try/catch/finally block, and in the same commit I updated the class to conform to DSpace 7 code style for easier porting.

kshepherd committed:
[DS-3940] try/catch for parsing before finally closing stream, d7 code style

kshepherd commented:
Merging as I have review and +1 from a committer, +1 from another committer (though not registered here as we work together) and 2 +1s from trusted devs who are members of release team.
Thanks all for the reviews and improvement suggestions.

kshepherd commented:
@benpal  yes, this should merge / cherry-pick cleanly into DSpace 6.3, and there is a DSpace CRIS 5.8 branch which I think should be fine to cherry-pick (see the last two commits) into standard DSpace 5.8, at https://github.com/the-library-code/DSpace/commits/SHERPA_API_v2_dspace-cris_5x

benpal commented:
Thank you so much sir @kshepherd.

benpal commented:
Hello everyone,

I'm having this error:
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project dspace-xmlui: Compilation failure
[ERROR] /build/dspace-6.3/dspace-xmlui/src/main/java/org/dspace/app/xmlui/aspect/submission/submit/UploadStep.java:[286,39] cannot find symbol
[ERROR]   symbol:   method getSizeBytes()
[ERROR]   location: variable bitstream of type org.dspace.content.Bitstream

kshepherd commented:
> Hello everyone,
> 
> I'm having this error:
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project dspace-xmlui: Compilation failure
> [ERROR] /build/dspace-6.3/dspace-xmlui/src/main/java/org/dspace/app/xmlui/aspect/submission/submit/UploadStep.java:[286,39] cannot find symbol
> [ERROR] symbol: method getSizeBytes()
> [ERROR] location: variable bitstream of type org.dspace.content.Bitstream

Hi @benpal -  this is an example of the kind of difference between DSpace 5.x and 6.x -- getSize() is the 5.x method and getSizeBytes() is for 6.x

It looks like you're building 6.x here so I'm not sure why it's complaining about this method... is it possible there is a mixup with some source? Pretty sure this method isn't called by any of the commits in this PR.

benpal commented:
> > Hello everyone,
> > I'm having this error:
> > [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project dspace-xmlui: Compilation failure
> > [ERROR] /build/dspace-6.3/dspace-xmlui/src/main/java/org/dspace/app/xmlui/aspect/submission/submit/UploadStep.java:[286,39] cannot find symbol
> > [ERROR] symbol: method getSizeBytes()
> > [ERROR] location: variable bitstream of type org.dspace.content.Bitstream
> 
> Hi @benpal - this is an example of the kind of difference between DSpace 5.x and 6.x -- getSize() is the 5.x method and getSizeBytes() is for 6.x
> 
> It looks like you're building 6.x here so I'm not sure why it's complaining about this method... is it possible there is a mixup with some source? Pretty sure this method isn't called by any of the commits in this PR.

Thank you @kshepherd, Yes I am testing both in 5.8 and 6.3, but for this case, I am using 6.3 as my test case. It seems that the method 'getSizeBytes()' symbol is not recognized by maven compiler.

eulereadgbe commented:
Hi @kshepherd. It seems the authority control for SHERPARoMEOPublisher and SHERPARoMEOJournalTitle are not working. I tested it both in JSPUI and XMLUI and using suggest and lookup in the choices.presentation. When using lookup, the error that pops up is "Failed to load choice data:  Server indicates error in response." The SHERPA/RoMEO Publisher Policy Database for JSPUI and the Publisher information for XMLUI during the Upload stage however are working correctly. Please advise on how to make the authority control for publisher and journal-title work. Thanks in advance!

kshepherd commented:
Hi @eulereadgbe - this issue was specifically about fixing the submission-time SHERPA lookups, not the authority plugins, so they'll need updating separately unfortunately. It should be not too difficult to refactor the plugins to use the new v2 classes for Publisher and Journal, but it's not something I've looked at yet.
